llvm/include/llvm/Support/SimpleTable.h

```diff
@@ -0,0 +1,143 @@
+
+    void setParent(SimpleTable *P) {
+      assert(Parent == nullptr && "parent already set");
+      Parent = P;
+    }
+
+  private:
+    friend class SimpleTable;
+
+    Row(SimpleTable *Parent) : Parent(Parent) {}
+
+    Row(SimpleTable *Parent, ArrayRef<StringRef> R) : Row(Parent) {
+      for (auto Cell : R)
+        Cells.emplace_back(Cell.str());
+    }
+
+    std::string &operator[](int I) { return Cells[I]; }
+
+    const std::string &operator[](int I) const { return Cells[I]; }
+
+  private:
+    std::vector<std::string> Cells;
+    SimpleTable *Parent;
+  };
+
+public:
+  SimpleTable() = default;
+  static Expected<UPtrTy> create(ArrayRef<StringRef> ColNames);
+  static Expected<UPtrTy> create(int NColumns);
+  int getNumColumns() const { return static_cast<int>(ColumnNames.size()); }
+  int getNumRows() const { return static_cast<int>(rows().size()); }
+
+  // Add a column with given title and assign cells to given values. The table
+  // must be empty or the number of the input cells must match column size.
+  Error addColumn(const Twine &Title, ArrayRef<StringRef> Cells);
+  Error addColumn(const Twine &Title, ArrayRef<std::string> Cells);
+
+  // Replaces a column in this table with another column of the same size from
+  // another table. Columns are identified by their names. If source column name
+  // is empty, it is assumed to match the source's name.
+  Error replaceColumn(StringRef Name, const SimpleTable &Src,
+                      StringRef SrcName = "");
+
+  // Renames a column.
+  Error renameColumn(StringRef OldName, StringRef NewName);
+
+  // Removes all columns except those with given names.
+  Error peelColumns(ArrayRef<StringRef> ColNames);
+
+  // Iterates all cells top-down lef-right and adds their values to given
+  // container.
+  void linearize(std::vector<std::string> &Res) const;
+
+  // Serialized the table to a stream.
+  void write(raw_ostream &Out, bool WriteTitles = true,
+             char ColSep = '|') const;
+
+  // De-serializes a table from a stream.
+  static Expected<UPtrTy> read(MemoryBuffer *Buf, char ColSep = '|');
+
+  // De-serializes a table from a file.
+  static Expected<UPtrTy> read(const Twine &FileName, char ColSep = '|');
+
+  const std::vector<Row> &rows() const { return Rows; }
+
+  void addRow(ArrayRef<StringRef> R) {
+    assert((R.size() == ColumnNames.size()) && "column number mismatch");
+    Rows.emplace_back(Row(this, R));
+  }
+
+  int getColumnId(StringRef ColName) const;
+
+  Row &operator[](int I) { return Rows[I]; }
+  const Row &operator[](int I) const { return Rows[I]; }
+
+private:
+  Error addColumnName(StringRef ColName);
+  void rebuildName2NumMapping();
+
+  std::map<StringRef, int> ColumnName2Num;
+  // Use list as the holder of string objects as modification never invalidate
+  // element addresses and iterators, unlike vector.
+  std::list<std::string> ColumnNames;
+  std::vector<std::list<std::string>::iterator> ColumnNum2Name;
+  std::vector<Row> Rows;
```

Comment:
I guess LLVM containers optimized for small # of elements could be useful here.

Reply:
ok
but this would be potentially dangerous change on cut-off date.